From e45f990f782424f516a327a361f4b1aae7259d97 Mon Sep 17 00:00:00 2001
From: Steve Best <sbest@redhat.com>
Date: Tue, 8 Jul 2014 17:28:31 -0400
Subject: [tools] power/turbostat: Run on Broadwell

Message-id: <20140708172831.20977.23168.sendpatchset@ibm-p7r2-01.lab.bos.redhat.com>
Patchwork-id: 85525
O-Subject: [PATCH RHEL7.1 BZ1086455] tools/power turbostat: Run on Broadwell
Bugzilla: 1086455
RH-Acked-by: Don Zickus <dzickus@redhat.com>
RH-Acked-by: Neil Horman <nhorman@redhat.com>

RHBZ#:
------
https://bugzilla.redhat.com/show_bug.cgi?id=1086455

Description:
------------

Signed-off-by: Len Brown <len.brown@intel.com>

RHEL Version Found:
-------------------
RHEL 7.0

Brew:
-----
https://brewweb.devel.redhat.com/taskinfo?taskID=7676257

Upstream:
---------
http://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/commit/?id=4e8e863fed2e82278d29c6357de8251adb73acb9

Test Status:
------------
Tested by me using intel-sharkbay-ult-03 which is CPU family 6 model 61.

before the patch

[root@intel-sharkbay-ult-03 ~]# turbostat
cor CPU    %c0  GHz  TSC SMI    %c1    %c3    %c6 CTMP PTMP   %pc3   %pc6
          0.03 1.79 1.60   0  99.97   0.00   0.00   26   26   0.00   0.00
  0   0   0.01 1.75 1.60   0  99.99   0.00   0.00   26   26   0.00   0.00
  0   2   0.00 1.74 1.60   0 100.00
  1   1   0.11 1.79 1.60   0  99.89   0.00   0.00   25
  1   3   0.01 1.75 1.60   0  99.99

after the patch is applied the additional information is displayed.

[root@intel-sharkbay-ult-03 ~]# turbostat
cor CPU    %c0  GHz  TSC SMI    %c1    %c3    %c6    %c7 CTMP PTMP   %pc2   %pc3   %pc6   %pc7   %pc8   %pc9  %pc10  Pkg_W  Cor_W GFX_W
          0.05 1.76 1.59   0   5.91   0.00   0.00  94.04   26   27  88.10   0.00   0.00   0.00   0.00   0.00   0.00   1.70   0.01  0.00
  0   0   0.03 1.76 1.59   0  11.68   0.00   0.00  88.29   26   27  88.10   0.00   0.00   0.00   0.00   0.00   0.00   1.70   0.01  0.00
  0   2   0.03 1.78 1.59   0  11.68
  1   1   0.10 1.75 1.59   0   0.10   0.00   0.00  99.80   25
  1   3   0.02 1.78 1.59   0   0.18

---------------------------------------------------------------
Steve Best

Proposed Patch:
---------------

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/power/x86/turbostat/turbostat.c b/tools/power/x86/turbostat/turbostat.c
index 72be53c..466f96e 100644
--- a/tools/power/x86/turbostat/turbostat.c
+++ b/tools/power/x86/turbostat/turbostat.c
@@ -1502,6 +1502,9 @@ int has_nehalem_turbo_ratio_limit(unsigned int family, unsigned int model)
  case 0x46: /* HSW */
  case 0x37: /* BYT */
  case 0x4D: /* AVN */
+ case 0x3D: /* BDW */
+ case 0x4F: /* BDX */
+ case 0x56: /* BDX-DE */
   return 1;
  case 0x2E: /* Nehalem-EX Xeon - Beckton */
  case 0x2F: /* Westmere-EX Xeon - Eagleton */
@@ -1615,6 +1618,7 @@ void rapl_probe(unsigned int family, unsigned int model)
  case 0x3C: /* HSW */
  case 0x45: /* HSW */
  case 0x46: /* HSW */
+ case 0x3D: /* BDW */
   do_rapl = RAPL_PKG | RAPL_CORES | RAPL_CORE_POLICY | RAPL_GFX | RAPL_PKG_POWER_INFO;
   break;
  case 0x3F: /* HSX */
@@ -1622,6 +1626,8 @@ void rapl_probe(unsigned int family, unsigned int model)
   break;
  case 0x2D:
  case 0x3E:
+ case 0x4F: /* BDX */
+ case 0x56: /* BDX-DE */
   do_rapl = RAPL_PKG | RAPL_CORES | RAPL_CORE_POLICY | RAPL_DRAM | RAPL_PKG_PERF_STATUS | RAPL_DRAM_PERF_STATUS | RAPL_PKG_POWER_INFO;
   break;
  case 0x37: /* BYT */
@@ -1861,6 +1867,9 @@ int is_snb(unsigned int family, unsigned int model)
  case 0x3F: /* HSW */
  case 0x45: /* HSW */
  case 0x46: /* HSW */
+ case 0x3D: /* BDW */
+ case 0x4F: /* BDX */
+ case 0x56: /* BDX-DE */
   return 1;
  }
  return 0;
@@ -1872,7 +1881,8 @@ int has_c8_c9_c10(unsigned int family, unsigned int model)
   return 0;
 
  switch (model) {
- case 0x45:
+ case 0x45: /* HSW */
+ case 0x3D: /* BDW */
   return 1;
  }
  return 0;
-- 
1.7.1